### DISCUSSION TITLE
Textual Development in PyCharm/IDEA products

### DISCUSSION THREAD

**Discussion by ciaransweet** [Textualize/textual#628] [Q&A]
Hey folks, 

Have a few questions around how people are developing Textual Apps in products like PyCharm.

If I use the example: https://github.com/Textualize/textual#creating-widgets and run my app via iTerm2:

https://user-images.githubusercontent.com/9111975/181754981-ae4206c0-c69b-4acf-b356-a79505f38235.mov

I can easily see the `on_enter` and `on_exit` behaviour.

However if I try and run this in the PyCharm Terminal:

https://user-images.githubusercontent.com/9111975/181755229-700352aa-e4a8-40f0-bb4f-743dc1c9f627.mov

You can see that the hover functionality isn't quite correct. I have to physically click **and** drag the cursor for it to register these events.

**Question 1** - Is there something obvious to change in PyCharm/IDEA IDEs to change this behaviour? Are other people experiencing this?

---

Also, similar situation, it'd be nice to be able to `Run` the application, with the usual `Run <>` options in PyCharm.

I get even stranger output on that view:

https://user-images.githubusercontent.com/9111975/181755885-c4de82a9-81a9-4311-892d-774813f9665e.mov

I fully appreciate that this is likely an issue with this being the python interpreter view so likely not supported. But it would be nice 😛 

**Question 2** - If people are actively dev-ing in PyCharm/IDEA IDEs, is there a nice run configuration I can set so that the `Run` option gives me a nice textual experience? 😄 

---

Appreciate any input folks have! ❤️

**Comment by willmcgugan** [comment]
Not a PyCharm user, but from the amount of issues we get, PyCharm doesn't emulate a terminal all that well. I think there was some checkbox in settings that enables terminal emulation which *may* improve things.

VSCode has a solid terminal emulator, but I tend to switch to iTerm to test Textual code.

**Reply by ciaransweet** [reply]
I figured this might be the case 🤣 

I ask because I'm looking at writing tools for Java Devs (😱) and switching out of the IDE would be a pain (I assume).

Still. Not end of the world!

**Comment by illeatmyhat** [comment]
I had a similar issue. In PyCharm's Run/Debug Configurations, there's a "Emulate terminal in output console" button that you can press and it works better.
I've not done an exhaustive test of features but at least the centered button tutorial works.
